contributors/design-proposals/customresources-validation.md

```diff
@@ -0,0 +1,493 @@
+                    }
+                }
+            }
+        }
+    }
+}
+```
+
+## Validation Behavior
+
+The schema will be described in the `CustomResourceDefinitionSpec`. The validation will be carried out using the [go-openapi validation library](https://github.com/go-openapi/validate).
+ 
+While creating/updating the CR, the metadata is first validated. To validate the CR against the spec in the CRD, we _must_ have server-side validation and we _can_ have client-side validation.
+
+### Metadata
+
+ObjectMeta and TypeMeta are implicitly specified. They do not have to be added to the JSON-Schema of a CRD. The validation already happens today as part of the apiextensions-apiserver REST handlers.
+
+### Server-Side Validation
+
+The server-side validation is carried out after sending the request to the apiextensions-apiserver, i.e. inside the CREATE and UPDATE handlers for CRs.
+
+We do a schema pass there using the https://github.com/go-openapi/validate validator with the provided schema in the corresponding CRD. Validation errors are returned to the caller as for native resources.
+
+JSON-Schema also allows us to reject additional fields that are not defined in the schema and only allow the fields that are specified. This can be achieved by using `"additionalProperties": false` in the schema.
+
+### Client-Side Validation
+
+The client-side validation is carried out before sending the request to the api-server, or even completely offline. This can be achieved while creating resources through the client i.e. kubectl using the --validate option.
+
+If the API type serves the JSON-Schema in the swagger spec, the existing kubectl code will already be able to also validate CRs. This will be achieved as a follow-up.
+
+### Comparison between server-side and client-side Validation
+
+The table below shows the cases when server-side and client-side validation methods are applicable.
+
+| Case                                               | Server-Side  | Client-Side   |
+|:--------------------------------------------------:|:------------:|:-------------:|
+| Kubectl create/edit/replace with validity feedback | &#x2713;     | &#x2713;      | 
+| Custom controller creates/updates CRs              | &#x2713;     | &#x2717;      |
+| CRs are created by an untrusted party              | &#x2713;     | &#x2717;      |
+| Not making validation for CRs a special case       | &#x2713;     | &#x2717;      |
+
+The above table is an evidence that we need server-side validation as well, next to the client-side validation we easily get, nearly for free, by serving Swagger/OpenAPI specs in apiextension-apiserver. 
+
+This is especially true in situations when CRs are used by components that are out of the control of the admin. Example: A user can create a database CR for a Database-As-A-Service. In this case, only server-side validation can give confidence that the CRs are well formed.
+
+### Existing Instances and changing the Schema
```

Comment:
In the Google Doc, there was some discussion about ways to deal with CRs that exist before the introduction of validation on a given CRD for the first time. Can you summarize the findings in this proposal?